pyiceberg/table/__init__.py

```diff
@@ -1086,6 +1094,78 @@ def name_mapping(self) -> Optional[NameMapping]:
         """Return the table's field-id NameMapping."""
         return self.metadata.name_mapping()
 
+    def upsert(
+        self, df: pa.Table, join_cols: list[str], when_matched_update_all: bool = True, when_not_matched_insert_all: bool = True
+    ) -> UpsertResult:
+        """Shorthand API for performing an upsert to an iceberg table.
```

Comment:
Hey everyone,

I've been following the project and found this PR really interesting! Regarding the point you mentioned, @mattmartin14, I pulled your fork and ran _make docs-build_ with these changes. Based on my tests, I believe these modifications will fix this step:

- Removing self from the Args section
- Indenting Example one level deeper

Like this:

```suggestion
        """Shorthand API for performing an upsert to an iceberg table.

Args:
    df: The input dataframe to upsert with the table's data.
    join_cols: The columns to join on. These are essentially analogous to primary keys
    when_matched_update_all: Bool indicating to update rows that are matched but require an update due to a value in a non-key column changing
    when_not_matched_insert_all: Bool indicating new rows to be inserted that do not match any existing rows in the table

Example:
    Case 1: Both Parameters = True (Full Upsert)
```

Reply:
> I saw the only thing that failed in the CI was the docs build. Any idea why and if it's something I can fix?

CI's running `make docs-build`, you can also run this locally.

```
WARNING -  griffe: pyiceberg/table/__init__.py:1105: No type or annotation for parameter 'self'
WARNING -  griffe: pyiceberg/table/__init__.py:1111: Parameter 'Example' does not appear in the function signature

Aborted with 2 warnings in strict mode!
```

we build docs in strict mode; there are issues with the current docstring as outlined above

Reply:
Thanks @marcoaanogueira , @kevinjqliu ,

I'll get those changes in later today and hopefully we are good to go.